torch/csrc/autograd/python_saved_variable_hooks.cpp

```diff
@@ -45,12 +45,12 @@ namespace torch { namespace autograd {
     }
   }
 
-  // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-  PyObject* PyDefaultSavedVariableHooks::pack_hook_ = nullptr;
-  // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
-  PyObject* PyDefaultSavedVariableHooks::unpack_hook_ = nullptr;
+  std::mutex PyDefaultSavedVariableHooks::mutex_;
+  PyObject* PyDefaultSavedVariableHooks::pack_hook_(nullptr);
+  PyObject* PyDefaultSavedVariableHooks::unpack_hook_(nullptr);
 
   void PyDefaultSavedVariableHooks::set_hooks(py::function &pack_hook, py::function &unpack_hook) {
+    std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
For this to be effective, you need to protect every function that changes these hooks so the reset below in particular.

Also if you don't protect the get_hook, you can still get in a situation where a user can get a pack hook from one pair and an unpack from another (or one empty). Which is bad as well.